File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/LoginCommand.java
package ro.razvan.java.dagger.atm;

import javax.inject.Inject;

public final class LoginCommand implements SingleArgCommand {

    public static final String COMMAND_KEY_LOGIN = "login";

    private final Outputter outputter;
    private final Database database;
    private final UserCommandsRouter.Factory userCommandsRouterFactory;

    @Inject
    public LoginCommand(
            Outputter outputter,
            Database database,
            UserCommandsRouter.Factory userCommandsRouterFactory
    ) {
        this.outputter = outputter;
        this.database = database;
        this.userCommandsRouterFactory = userCommandsRouterFactory;
    }

    @Override
    public Result handleArg(String username) {
        final var account = database.getAccount(username);
        outputter.output(account.username() + " is logged in with balance: " + account.balance());
        return Result.enterNestedCommandSet(userCommandsRouterFactory.create(account).router());
    }

    @Override
    public String key() {
        return COMMAND_KEY_LOGIN;
    }
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/LoginCommand.kt
package ro.razvan.kotlin.dagger.atm

import ro.razvan.kotlin.dagger.atm.Command.Result
import java.util.*
import javax.inject.Inject

const val COMMAND_KEY_LOGIN = "login"

class LoginCommand @Inject constructor(
    private val outputter: Outputter,
    private val database: Database,
    private val userCommandsRouterFactory: UserCommandsRouter.Factory,
    private val account: Optional<Database.Account>
) : SingleArgCommand {

    override val key: String
        get() = COMMAND_KEY_LOGIN

    override fun handleArg(arg: String): Result {
        if (account.isPresent) {
            outputter("${account.get().username} is already logged in!")
            return Result.handled()
        }
        val account = database.getAccount(username = arg)
        outputter("$arg is logged in with balance: ${account.balance}")
        return Result.enterNestedCommandSet(router = userCommandsRouterFactory.create(account).router())
    }
}